lib/trie/database.go

```diff
@@ -185,22 +185,28 @@ func (t *Trie) loadNode(db Database, n *Node) error {
 	return nil
 }
 
-// PopulateMerkleValues writes the Merkle value of each children of the node given
-// as keys to the map merkleValues.
-func (t *Trie) PopulateMerkleValues(n *Node, merkleValues map[string]struct{}) {
-	if n.Kind() != node.Branch {
+// PopulateMerkleValues writes the Merkle values of the node given and of
+// all its descendant nodes as keys to the map merkleValues.
+// It is assumed the node and its descendant nodes have their Merkle value already
+// computed.
+func PopulateMerkleValues(n *Node, merkleValues map[string]struct{}) {
+	if n == nil {
 		return
 	}
 
-	branch := n
-	for _, child := range branch.Children {
-		if child == nil {
-			continue
-		}
+	if len(n.MerkleValue) == 0 {
+		panic(fmt.Sprintf("node with key 0x%x has no Merkle value computed", n.Key))
```

Comment:
We do want to panic, it would be a programming error (and catastrophic consequences) if this ever happens. This is essentially a guard against my changes on other branches based on this one (since it's a lot of mental overhead to think about everything, the trie code is quite hard). This will be removed soon enough with lazy loading since all Merkle values will be 100% computed anyway, so that panic will be removed.
I'll add a `// TODO` comment linking the lazy loading issue.